In [1]:
import pandas as pd
import numpy as np
import tensorflow as tf
import sklearn

In [2]:
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense
from tensorflow.keras import utils as np_utils
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix

In [10]:
base = pd.read_csv('iris_aula.csv')

In [14]:
base

sepal length  sepal width  petal length  petal width           class
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

[150 rows x 5 columns]

In [20]:
X = base.iloc[:, :4].values
X

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [63]:
y = base.iloc[:, 4].values
y

array(['Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',
       'Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',
       'Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',
       'Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',
       'Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',
       'Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',
       'Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',
       'Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',
       'Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',
       'Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',
       'Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',
       'Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',
       'Iris-setosa', 'Iris-setosa', 'Iris-versicolor', 'Iris-versicolor',
       'Iris-versicolor', 'Iris-versicolor', 'Iris-versicolor',
       'Iris-versicolor', 'Iris-versicolor', 'Iris-versic

#### setosa -> 0
#### versicolor -> 1
#### virginica -> 2
                 

In [67]:
label_encoder = LabelEncoder()
y =  label_encoder.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

Transformação em uma forma que a rede neural consiga reproduzir, e transformação em um array de 150 exemplos e 3 categorias

In [69]:
y = np_utils.to_categorical(y)
y

array([[1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0

Criação dos dados de teste e treinamento

In [76]:
X_train, X_test, y_train, y_test =  train_test_split(X, y, test_size = 0.25)

In [78]:
X_train.shape, X_test.shape

((112, 4), (38, 4))

In [82]:
y_train.shape, y_test.shape

((112, 3), (38, 3))

Criação da rede neural

In [86]:
rede_neural = Sequential([
    tf.keras.layers.InputLayer(shape = (4,)),
    tf.keras.layers.Dense(units = 4, activation = 'relu'),
    tf.keras.layers.Dense(units = 4, activation = 'relu'),
    tf.keras.layers.Dense(units = 3, activation = 'softmax')    
])

In [88]:
rede_neural.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 4)                   │              20 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 4)                   │              20 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 3)                   │              15 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 55 (220.00 B)

 Trainable params: 55 (220.00 B)

 Non-trainable params: 0 (0.00 B)

In [98]:
rede_neural.compile(optimizer = 'adam', loss = 'categorical_crossentropy', metrics = ['categorical_accuracy'])

In [101]:
rede_neural.fit(X_train, y_train, batch_size = 10, epochs = 500)

Epoch 1/500
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - categorical_accuracy: 0.9710 - loss: 0.0475 
Epoch 2/500
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - categorical_accuracy: 0.9872 - loss: 0.0566 
Epoch 3/500
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - categorical_accuracy: 0.9781 - loss: 0.0476 
Epoch 4/500
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - categorical_accuracy: 0.9832 - loss: 0.0689 
Epoch 5/500
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - categorical_accuracy: 0.9846 - loss: 0.0344  
Epoch 6/500
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - categorical_accuracy: 0.9820 - loss: 0.0440 
Epoch 7/500
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - categorical_accuracy: 0.9679 - loss: 0.0670 
Epoch 8/500
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - categorical_accuracy: 0.9712 - loss: 0.0544 
Epoch 9/500
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - categorical_accuracy: 0.9760 - loss: 0.0653  
Epoch 10/500
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - categorical_accuracy: 0.9925 - loss: 0.0487 
Epoch 11/500
12/1

No treinamento ele teve um accuracy de 98% e um loss de 3%, já na base de test ele teve um accuracy de 95% e um loss de 5%

In [103]:
rede_neural.evaluate(X_test, y_test)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - categorical_accuracy: 0.9545 - loss: 0.0507 


[0.05691073462367058, 0.9473684430122375]

In [110]:
previsoes = rede_neural.predict(X_test)
print(previsoes)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step
[[3.43618603e-06 1.57283336e-01 8.42713237e-01]
 [1.23059493e-04 9.94296491e-01 5.58050396e-03]
 [8.28206947e-11 3.35339246e-05 9.99966502e-01]
 [1.00000000e+00 1.20744073e-08 5.16023890e-09]
 [1.23059493e-04 9.94296491e-01 5.58050396e-03]
 [1.00000000e+00 2.04748662e-09 1.15139831e-09]
 [2.73285786e-11 1.38394835e-05 9.99986172e-01]
 [1.23059493e-04 9.94296491e-01 5.58050396e-03]
 [1.00000000e+00 1.90859484e-09 1.10509668e-09]
 [1.00000000e+00 2.02547742e-10 1.74437340e-10]
 [1.00000000e+00 2.29406183e-08 9.24580501e-09]
 [6.44060971e-10 1.72396569e-04 9.99827623e-01]
 [9.09341491e-09 1.42635254e-03 9.98573661e-01]
 [1.00000000e+00 4.07959194e-10 3.10940301e-10]
 [1.23059493e-04 9.94296491e-01 5.58050396e-03]
 [9.09407390e-05 9.82504308e-01 1.74047593e-02]
 [1.00000000e+00 3.65275064e-08 1.48673918e-08]
 [1.00000000e+00 9.70778666e-11 9.10252290e-11]
 [7.54574867e-05 9.66373146e-01 3.35513577e-02]
 [1.21849909e-04 9.94081438e-01 5.79667836e-03]
 [

In [114]:
previsoes = previsoes > 0.5
print(previsoes)

[[False False  True]
 [False  True False]
 [False False  True]
 [ True False False]
 [False  True False]
 [ True False False]
 [False False  True]
 [False  True False]
 [ True False False]
 [ True False False]
 [ True False False]
 [False False  True]
 [False False  True]
 [ True False False]
 [False  True False]
 [False  True False]
 [ True False False]
 [ True False False]
 [False  True False]
 [False  True False]
 [ True False False]
 [False False  True]
 [ True False False]
 [False  True False]
 [False  True False]
 [ True False False]
 [False  True False]
 [False False  True]
 [False  True False]
 [ True False False]
 [ True False False]
 [ True False False]
 [False False  True]
 [False False  True]
 [ True False False]
 [False False  True]
 [False  True False]
 [False  True False]]


Transformamos os dados de teste em suas respectivas iris, como ja foi passada.

In [118]:
y_test2 = [np.argmax(t) for t in y_test]
print(y_test2)

[2, 1, 2, 0, 1, 0, 2, 1, 0, 0, 0, 2, 2, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 2, 1, 0, 0, 0, 1, 2, 0, 2, 1, 1]


Transformamos os dados de previsoes em suas respectivas iris, como ja foi passada.

In [124]:
previsoes2 = [np.argmax(t) for t in previsoes]
previsoes2

[2,
 1,
 2,
 0,
 1,
 0,
 2,
 1,
 0,
 0,
 0,
 2,
 2,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 2,
 0,
 1,
 1,
 0,
 1,
 2,
 1,
 0,
 0,
 0,
 2,
 2,
 0,
 2,
 1,
 1]

In [133]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test2, previsoes2)

0.9473684210526315

In [135]:
confusion_matrix(y_test2, previsoes2)

array([[15,  0,  0],
       [ 0, 13,  2],
       [ 0,  0,  8]], dtype=int64)